packages/phone-number-privacy/combiner/src/match-making/get-contact-matches.ts

```diff
@@ -36,16 +43,113 @@ export async function handleGetContactMatches(
       return
     }
 
-    const { account, userPhoneNumber, contactPhoneNumbers, hashedPhoneNumber } = request.body
+    const {
+      account,
+      userPhoneNumber,
+      contactPhoneNumbers,
+      hashedPhoneNumber,
+      signedUserPhoneNumber,
+    } = request.body
 
     if (!(await isVerified(account, hashedPhoneNumber, getContractKit(), logger))) {
       respondWithError(response, 403, WarningMessage.UNVERIFIED_USER_ATTEMPT_TO_MATCHMAKE, logger)
       return
     }
+
+    // Verify that signedUserPhoneNumber was signed by the account's DEK
+    let signedUserPhoneNumberHash: string | undefined
+    if (signedUserPhoneNumber) {
+      let isValidSig = false
+      let didRetrieveDEK = true
+      try {
+        isValidSig = await verifySignedUserPhoneNumberSignature(
+          account,
+          userPhoneNumber,
+          signedUserPhoneNumber,
+          logger
+        )
+      } catch (error) {
+        if (error.message === ErrorMessage.CONTRACT_GET_FAILURE) {
+          logger.warn(
+            'Failed to retrieve DEK to verify signedUserPhoneNumber. Signature is assumed valid and request is not recorded.'
+          )
+          isValidSig = true
+          didRetrieveDEK = false
+        } else {
+          throw error
+        }
+      }
+      if (!isValidSig) {
+        respondWithError(response, 403, WarningMessage.INVALID_USER_PHONE_NUMBER_SIGNATURE, logger)
+        return
+      }
+      if (didRetrieveDEK) {
+        // Hash signed number for good measure
```

Comment:
Generally, ECDSA signatures are not deterministic, so if you store the signature itself, it is possible for a correct client to come along later and provide a valid distinct signature and have this check fail through no fault of their own. If your goal is to bind the DEK used here, I would recommend storing the DEK public key or a hash of the public key (e.g. the Celo address) instead.

Reply:
Ok, nevermind. I see what is happening here now. The goal is to bind an account to a single phone number (hash) without linking the two in our databases. It's definitly a bit funky, and relies on non-standard behavior, but this does make sense and does work 👍